File: Packages/Learning/Sources/Learning/LearningEvent.swift
import Foundation

/// Per-word language attribution for learned counts.
///
/// This is attribution metadata only. Storage is SURFACE FORMS — the hint
/// says "this commit happened while the engine believed the lane was
/// Icelandic/English", it never triggers any lemma-level merging (see
/// `LemmaBoostProviding` for the wave-2 seam).
public enum LanguageHint: String, Codable, CaseIterable, Sendable {
    case icelandic = "is"
    case english = "en"
    case unknown = "un"
}

/// A single learning event, as emitted by the keyboard extension.
///
/// ## Privacy invariants (HARD — see also `LearningPrivacy`)
///
/// 1. **Single words / word pairs only. NEVER running text.** The largest
///    unit of language any event may carry is one word plus its immediate
///    predecessor (`wordCommitted.previousWord`). No sentence, no context
///    window, no n>2 grams.
/// 2. **No timestamps finer than a coarse day bucket.** Events are stamped
///    with a UTC day number (`DayBucket`) at append time — never a
///    time-of-day, never a monotonic clock.
/// 3. **Secure / URL / email / password fields must never be logged.** The
///    package cannot see the host text field, so this is enforced by the
///    caller (the extension's input pipeline). Call
///    `LearningPrivacy.assertLoggableFieldContext(...)` at the call boundary;
///    it traps in debug builds and invokes the installable
///    `violationHandler` in release builds.
public enum LearningEvent: Equatable, Sendable {
    /// A word was committed to the document (space/delimiter/return), with
    /// the immediately preceding committed word if one exists, and the
    /// engine's language-lane belief at commit time.
    case wordCommitted(word: String, previousWord: String?, languageHint: LanguageHint)
    /// The user tapped a suggestion-bar candidate: `typed` is the raw token,
    /// `accepted` is the candidate that was committed instead.
    case suggestionAccepted(typed: String, accepted: String)
    /// The user reverted an applied autocorrection: `original` is what they
    /// actually typed (and got back), `applied` is the correction they
    /// rejected.
    case correctionReverted(original: String, applied: String)
    /// The user tapped the quoted verbatim slot — the strongest explicit
    /// "this is a real word" signal (learns immediately, no day threshold).
    case wordTapped(word: String)
    /// One touch sample for the per-key adaptive touch model: `dx`/`dy` are
    /// offsets from the resolved key's center, normalized to key units
    /// (1.0 = one key width/height). Only confirmed text should train this
    /// model — the caller must emit samples for committed words only.
    case touchSample(keyChar: Character, dx: Double, dy: Double)
}

/// An event as read back from the log: the event plus the coarse UTC day
/// bucket it was appended on (the only temporal information ever stored).
public struct LoggedEvent: Equatable, Sendable {
    /// UTC day bucket (days since 1970-01-01, `DayBucket`).
    public let day: Int32
    public let event: LearningEvent

    public init(day: Int32, event: LearningEvent) {
        self.day = day
        self.event = event
    }
}

/// Coarse day bucketing — the maximum temporal resolution this package is
/// allowed to persist (privacy invariant #2 on `LearningEvent`).
public enum DayBucket {
    /// Whole days since the Unix epoch, UTC. Deliberately not calendar-aware:
    /// the only property compaction needs is "distinct buckets ≈ distinct
    /// days", and a fixed 86 400 s bucket has no time-zone leakage.
    public static func bucket(for date: Date) -> Int32 {
        Int32(floor(date.timeIntervalSince1970 / 86_400))
    }

    public static func current() -> Int32 {
        bucket(for: Date())
    }
}

File: AndroidClient/lib/engine/src/main/kotlin/is/solberg/lyklabord/engine/learning/LearningEvent.kt
package `is`.solberg.lyklabord.engine.learning

import java.time.Instant
import kotlin.math.floor

/** Per-word language attribution for learned counts. */
enum class LanguageHint(val rawValue: String) {
    ICELANDIC("is"), ENGLISH("en"), UNKNOWN("un");
    companion object {
        val icelandic: LanguageHint get() = ICELANDIC
        val english: LanguageHint get() = ENGLISH
        val unknown: LanguageHint get() = UNKNOWN
    }
}

/** A single privacy-bounded learning event. */
sealed interface LearningEvent {
    companion object {
        fun wordCommitted(word: String, previousWord: String? = null, languageHint: LanguageHint = LanguageHint.UNKNOWN) =
            WordCommitted(word, previousWord, languageHint)
        fun suggestionAccepted(typed: String, accepted: String) = SuggestionAccepted(typed, accepted)
        fun correctionReverted(original: String, applied: String) = CorrectionReverted(original, applied)
        fun wordTapped(word: String) = WordTapped(word)
        fun touchSample(keyChar: Char, dx: Double, dy: Double) = TouchSample(keyChar, dx, dy)
    }
    data class WordCommitted(
        val word: String,
        val previousWord: String? = null,
        val languageHint: LanguageHint = LanguageHint.UNKNOWN,
    ) : LearningEvent

    data class SuggestionAccepted(val typed: String, val accepted: String) : LearningEvent
    data class CorrectionReverted(val original: String, val applied: String) : LearningEvent
    data class WordTapped(val word: String) : LearningEvent
    data class TouchSample(val keyChar: Char, val dx: Double, val dy: Double) : LearningEvent
}

data class LoggedEvent(val day: Int, val event: LearningEvent)

object DayBucket {
    fun bucket(forDate: Instant): Int = floor(forDate.epochSecond / 86_400.0).toInt()
    fun bucket(forDate: java.util.Date): Int = floor(forDate.time / 86_400_000.0).toInt()
    fun current(): Int = bucket(Instant.now())
}
